File: ident-uuid/src/main/scala/net/nicktelford/ident/uuid/rfc4122/NameUUID.scala
package net.nicktelford.ident.uuid.rfc4122

import java.security.MessageDigest
import net.nicktelford.ident.uuid._

/** Companion object for generating NameUUIDs. */
object NameUUID {

  /** Generates a NameUUID using the MD5 hashing algorithm. */
  object md5 extends NameUUIDFactory(MD5Scheme)

  /** Generates a NameUUID using the SHA1 hashing algorithm. */
  object sha1 extends NameUUIDFactory(SHA1Scheme)

  /** Generates a NameUUID using the default hashing algorithm. */
  def apply = sha1

  /** Base implementation for a NameUUID hashing scheme. */
  case class Scheme private[NameUUID] (version: Short, algorithm: String) {
    def hash(data: Iterable[Byte]): Iterable[Byte] = {
      MessageDigest.getInstance(algorithm).digest(data.toArray)
    }
  }

  /** MD5 NameUUID hashing scheme */
  case object MD5Scheme extends Scheme(3, "MD5")

  /** SHA1 NameUUID hashing scheme */
  case object SHA1Scheme extends Scheme(5, "SHA1")

  /** Factory for NameUUIDs. */
  class NameUUIDFactory private[NameUUID] (scheme: Scheme)
    extends RFC4122Factory[NameUUID] {

    /** @see UUIDFactory.MaxValue */
    val MaxValue: NameUUID = new NameUUID(encodeMSB(-1), encodeLSB(-1, -1), scheme)

    /** @see UUIDFactory.MinValue */
    val MinValue: NameUUID = new NameUUID(encodeMSB(0), encodeLSB(0, 0), scheme)

    /** @see RFC4122.version */
    val version: Short = scheme.version

    /** Generates a UUID for the given name using the given Scheme. */
    def apply(namespace: UUID, name: Iterable[Byte]): NameUUID = {
      val hash = scheme.hash(namespace.bytes ++ name)

      val timestampBits = bytesToLong(hash)
      val clockBits = bytesToInt(hash.slice(8, 12))
      val nodeBits = bytesToLong(hash.slice(12, 16))

      val msb = encodeMSB(timestampBits)
      val lsb = encodeLSB(clockBits, nodeBits)

      new NameUUID(msb, lsb, scheme)
    }

    def apply(msb: Long, lsb: Long): NameUUID = {
      NameUUID(msb, lsb, scheme)
    }

    // TODO: using a fold isn't very performant, although it is expressive
    /** Generates a Long from the first 8 bytes of a byte array. */
    private def bytesToLong(bytes: Iterable[Byte]): Long = {
      bytes.take(8).zipWithIndex.foldLeft(0L) {
        case (acc, (b, i)) => acc | (b << (7 - i) * 8)
      }
    }

    // TODO: using a fold isn't very performant, although it is expressive
    /** Generates an Int from the first 4 bytes of a byte array. */
    private def bytesToInt(bytes: Iterable[Byte]): Int = {
      bytes.take(4).zipWithIndex.foldLeft(0) {
        case (acc, (b, i)) => acc | (b << (7 - i) * 8)
      }
    }
  }

}


/** A Version 3 or 5, name-based UUID hashed from a namespace and name. */
case class NameUUID(msb: Long, lsb: Long, scheme: NameUUID.Scheme)
  extends UUID with RFC4122 {

  val version: Short = scheme.version
}

File: ident-java/src/main/java/net/nicktelford/ident/rfc4122/NameUUID.java
package net.nicktelford.ident.rfc4122;

import net.nicktelford.ident.UUID;
import net.nicktelford.ident.util.Bytes;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * An RFC4122 Version 3 or 5 Name-based UUID.
 * <p/>
 * The {@link #version()} of this UUID depends on the hashing algorithm chosen:
 * <ul>
 *     <li><b>3</b>: {@link #MD5}</li>
 *     <li><b>5</b>: {@link #SHA1}</li>
 * </ul>
 * To create a name-based UUID, first select an algorithm using one of the
 * builders; {@link #MD5} and {@link #SHA1}.
 * <p/>
 * It is nearly always preferable to use a {@link #SHA1} (version 5) UUID as
 * they have a lower probability of collisions than {@link #MD5} (version 3)
 * UUIDs.
 *
 * @see <a href="http://www.ietf.org/rfc/rfc4122">RFC4122</a>
 */
public class NameUUID extends RFC4122UUID {

    public static Scheme MD5 = Scheme.MD5;
    public static Scheme SHA1 = Scheme.SHA1;

    static enum Scheme {

        MD5(3, "md5"),
        SHA1(5, "sha1");

        private final short version;
        private final MessageDigest algorithm;
        private final Factory factory;

        public final NameUUID MAX_VALUE;
        public final NameUUID MIN_VALUE;

        Scheme(final int version,
               final String algorithm) throws NoSuchAlgorithmException {
            this((short) version, algorithm);
        }

        Scheme(final short version,
               final String algorithm) throws NoSuchAlgorithmException {
            this.version = (short) version;
            this.algorithm = MessageDigest.getInstance(algorithm);
            this.factory = new Factory(this.version);
            MAX_VALUE = new NameUUID(factory.encodeMSB(-1),
                                     factory.encodeLSB(-1),
                                     version);
            MIN_VALUE = new NameUUID(factory.encodeMSB(0),
                                     factory.encodeLSB(0),
                                     version);
        }

        public short version() {
            return version;
        }

        public NameUUID forName(final UUID namespace, final byte[] name) {
            final byte[] hash = hash(namespace.toBytes(), name);
            final long msb = factory.encodeMSB(Bytes.toLong(hash));
            final long lsb = factory.encodeLSB(Bytes.toLong(hash, 8));

            return new NameUUID(msb, lsb, version());
        }

        private byte[] hash(final byte[] data) {
            return algorithm.digest(data);
        }

        private byte[] hash(final byte[] namespace, final byte[] name) {
            return hash(Bytes.concat(namespace, name));
        }
    }

    private final short version;

    NameUUID(final long msb, final long lsb, final short version) {
        super(msb, lsb);
        this.version = version;
    }

    @Override
    public short version() {
        return version;
    }
}
